tests/e2e/bulk-pubsub/bulk_pubsub_test.go

```diff
@@ -0,0 +1,412 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bulkpubsubapp_e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"math/rand"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/ratelimit"
+
+	"github.com/dapr/dapr/tests/e2e/utils"
+	kube "github.com/dapr/dapr/tests/platforms/kubernetes"
+	"github.com/dapr/dapr/tests/runner"
+)
+
+var tr *runner.TestRunner
+
+const (
+	// Number of get calls before starting tests.
+	numHealthChecks = 60
+
+	// used as the exclusive max of a random number that is used as a suffix to the first message sent.  Each subsequent message gets this number+1.
+	// This is random so the first message name is not the same every time.
+	randomOffsetMax           = 49
+	numberOfMessagesToPublish = 60
+	publishRateLimitRPS       = 25
+
+	receiveMessageRetries = 5
+
+	publisherAppName      = "pubsub-publisher-bulk-subscribe"
+	bulkSubscriberAppName = "pubsub-bulk-subscriber"
+	PubSubEnvVar          = "DAPR_TEST_PUBSUB_NAME"
+)
+
+var (
+	offset     int
+	pubsubName string
+)
+
+// sent to the publisher app, which will publish data to dapr.
+type publishCommand struct {
+	ReqID       string            `json:"reqID"`
+	ContentType string            `json:"contentType"`
+	Topic       string            `json:"topic"`
+	Data        interface{}       `json:"data"`
+	Protocol    string            `json:"protocol"`
+	Metadata    map[string]string `json:"metadata"`
+}
+
+type callSubscriberMethodRequest struct {
+	ReqID     string `json:"reqID"`
+	RemoteApp string `json:"remoteApp"`
+	Protocol  string `json:"protocol"`
+	Method    string `json:"method"`
+}
+
+// data returned from the subscriber app.
+type receivedBulkMessagesResponse struct {
+	ReceivedByTopicRawSub     []string `json:"pubsub-raw-sub-topic"`
+	ReceivedByTopicCESub      []string `json:"pubsub-ce-sub-topic"`
+	ReceivedByTopicRawBulkSub []string `json:"pubsub-raw-bulk-sub-topic"`
+	ReceivedByTopicCEBulkSub  []string `json:"pubsub-ce-bulk-sub-topic"`
+}
+
+type cloudEvent struct {
+	ID              string `json:"id"`
+	Type            string `json:"type"`
+	DataContentType string `json:"datacontenttype"`
+	Data            string `json:"data"`
+}
+
+// checks is publishing is working.
+func publishHealthCheck(publisherExternalURL string) error {
+	commandBody := publishCommand{
+		ContentType: "application/json",
+		Topic:       "pubsub-healthcheck-topic-http",
+		Protocol:    "http",
+		Data:        "health check",
+	}
+
+	// this is the publish app's endpoint, not a dapr endpoint
+	url := fmt.Sprintf("http://%s/tests/publish", publisherExternalURL)
+
+	return backoff.Retry(func() error {
```

Comment:
Why is there a retry? I don't think we should retry in tests - if it doesn't work, that should be a bug right?

Reply:
I had used the same code for publisher health check, that was used for testing another App.
I think the motive here is to check if publisher is able to post messages successfully. And, without retry we would also take into account any network glitch etc. as well, whereas retry may be better in such cases, eliminating false negatives to a good extent.
I can still remove it, if the above logic doesn't go well.

Reply:
For the initial health checks, a retry is fine.

For publishing, we should rely on resiliency and the built-in retries, and not add more retries in our tests.